### PR TITLE
Added `MartenCommandNotSupportedException` to give more meaningful information if feature is not supported (eg. FTS)

### PR DESCRIPTION
**Pull request by oskardudycz**
This is the first draft of the Exceptions mapping to get the yays or nays for the approach.

@mysticmind do not merge it yet 🤣

### CODE REVIEW

**Comment by jokokko** [comment]
Looking good. Just thought, whether the enum is overkill? Or just the string reason would do. The current comparison logic would also trip up on something like `session.Query<object>("select tsvector(0)")`. Whether we consider it a problem is another thing.

**Comment by oskardudycz** [comment]
@joko thank you for the review 👍 
Regarding your concerns: I think that we’ll have more possible features incompatibility, we’re supporting PG from version 9, while soon there will be 12. For FTS we need also one more flag, as some features are supported from the version 11. I agree that it might look like overkill for now but imho it’s basic concept with preparation to evolve.
Regarding the check I’m also checking error code and the matching is done when exception was raised.
Does this answer clears your concerns? :)

**Comment by jokokko** [comment]
@oskardudycz Hello. What I meant with the `session.Query<object>("select tsvector(0)")` example is that it would throw `Marten Command Failure:Full Text Search needs at least Postgres version 10.`, which in this example would not explain the actual error case. I don't see that as a huge problem though... But I might have some idea how to tighten down `TryToMapToMartenCommandNotSupportedException` so it won't end up growing and growing over time.

Overall, looking good.

**Comment by oskardudycz** [comment]
@jokokko thank you for your feedback. I can think about that `session.Query<object>("select tsvector(0)")` but I also don't see it as a huge burden.
Regarding making `TryToMapToMartenCommandNotSupportedException` I was thinking about some pluggable ExceptionFilters (so simmilar as it's in ASP.NET), but I'm not sure if that won't be overengineering at this stage. What do you think?

File: src/Marten/Exceptions/MartenCommandExceptionFactory.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using Marten.Services;
+using Npgsql;
+
+namespace Marten.Exceptions
+{
+    /// <summary>
+    /// Class responsible for MartenCommandException creation.
+    /// Based on the exact command code and exception creates
+    /// generi MartenCommandException or specific Exception derived from it.
+    /// </summary>
```
Comment (mysticmind) [inline src/Marten/Exceptions/MartenCommandExceptionFactory.cs]: few tweaks as below:
Class responsible for creating MartenCommandException exception or exceptions derived from it based on exact command code

**Comment by jokokko** [comment]
So my PR #1288 would slightly modify the checks & add unit tests for this.

File: src/Marten/Exceptions/MartenCommandException.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            NpgsqlCommand command,
+            string prefix = null
+        )
+        {
+            if (prefix != null)
+            {
+                prefix = $"{prefix}\n";
+            }
+
+            return $"Marten Command Failure:\n{prefix}{command.CommandText}\n\n";
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply src/Marten/Exceptions/MartenCommandException.cs]: Fixed.

File: src/Marten/Exceptions/MartenCommandException.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+using System;
+using Npgsql;
+
+namespace Marten
+{
+    public class MartenCommandException: Exception
```
[the remark this answers is not in the document]
Comment (oskardudycz) [reply src/Marten/Exceptions/MartenCommandException.cs]: Added.

File: src/Marten/Exceptions/MartenCommandExceptionFactory.cs (hunk 1 above)
Comment (oskardudycz) [reply src/Marten/Exceptions/MartenCommandExceptionFactory.cs]: Updated.

**Comment by oskardudycz** [comment]
@jokokko I cherry picked manually your commits from https://github.com/JasperFx/marten/pull/1288. That was the fastest way for me to fix the conflicts ;) I renamed `KnownExceptionCause` to `KnownNotSupportedExceptionCause` to be more precise, as it's strictly related with `MartenCommandNotSupportedException`. I also made it Internal, as imho it's our internal implementation detail and it's better to keep something internal if it's not needed to be public. Then we have easier refactoring without API breaking changes :)

@mysticmind I applied your review comments.

I also added `NotSupportedReason.WebStyleSearchNeedsAtLeastPostgresVersion11` and test for that.

**Comment by oskardudycz** [comment]
@mysticmind that's good comment. I did a bit different:
- fixed the namespace on the `Exceptions/MartenCommandException.cs`
- added again MartenCommandException in the previous location, derived it from `Exceptions/MartenCommandException.cs` and marked as obsolete.

Thanks for that users will know that they need to move to the new one from the warning in VS after upgrade.
See: https://github.com/JasperFx/marten/pull/1286/commits/89c36cf580adf87b7b424f94c70b5a07b3606434

Thoughts?

**Comment by mysticmind** [comment]
@oskardudycz that is a clever fix, like that 👍

**Comment by mysticmind** [comment]
@oskardudycz failing unit test in PG9.6 ==> `Marten.Testing.Exceptions.known_exception_causes_dueto_pg10.can_map_web_style_search_not_supported [FAIL]`

**Comment by oskardudycz** [comment]
@mysticmind https://github.com/JasperFx/marten/pull/1286/commits/7a39fbd61cba584a21d049559c7d5baeb5b06fed that should solve it 👍

**Pull request merged**